// @strands-agents/sdk/dist/src/mcp.js
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { McpTool } from './tools/mcp-tool.js';
/** MCP Client for interacting with Model Context Protocol servers. */
export class McpClient {
    _clientName;
    _clientVersion;
    _transport;
    _connected;
    _client;
    constructor(args) {
        this._clientName = args.applicationName || 'strands-agents-ts-sdk';
        this._clientVersion = args.applicationVersion || '0.0.1';
        this._transport = args.transport;
        this._connected = false;
        this._client = new Client({
            name: this._clientName,
            version: this._clientVersion,
        });
    }
    get client() {
        return this._client;
    }
    /**
     * Connects the MCP client to the server.
     *
     * This function is exposed to allow consumers to connect manually, but will be called lazily before any operations that require a connection.
     *
     * @returns A promise that resolves when the connection is established.
     */
    async connect(reconnect = false) {
        if (this._connected && !reconnect) {
            return;
        }
        if (this._connected && reconnect) {
            await this._client.close();
            this._connected = false;
        }
        await this._client.connect(this._transport);
        this._connected = true;
    }
    /**
     * Disconnects the MCP client from the server and cleans up resources.
     *
     * @returns A promise that resolves when the disconnection is complete.
     */
    async disconnect() {
        // Must be done sequentially
        await this._client.close();
        await this._transport.close();
        this._connected = false;
    }
    /**
     * Lists the tools available on the server and returns them as executable McpTool instances.
     *
     * @returns A promise that resolves with an array of McpTool instances.
     */
    async listTools() {
        await this.connect();
        const result = await this._client.listTools();
        // Map the tool specifications to fully functional McpTool instances
        return result.tools.map((toolSpec) => {
            return new McpTool({
                name: toolSpec.name,
                description: toolSpec.description ?? '',
                inputSchema: toolSpec.inputSchema,
                client: this,
            });
        });
    }
    /**
     * Invoke a tool on the connected MCP server using an McpTool instance.
     * @param tool - The McpTool instance to invoke.
     * @param args - The arguments to pass to the tool.
     * @returns A promise that resolves with the result of the tool invocation.
     */
    async callTool(tool, args) {
        await this.connect();
        if (args === null || args === undefined) {
            return await this.callTool(tool, {});
        }
        if (typeof args !== 'object' || Array.isArray(args)) {
            throw new Error(`MCP Protocol Error: Tool arguments must be a JSON Object (named parameters). Received: ${Array.isArray(args) ? 'Array' : typeof args}`);
        }
        const result = await this._client.callTool({
            name: tool.name,
            arguments: args,
        });
        return result;
    }
}
