// @strands-agents/sdk/dist/src/__fixtures__/model-test-helpers.js
/**
 * Test fixtures and helpers for Model testing.
 * This module provides utilities for testing Model implementations without
 * requiring actual API clients.
 */
import { Model } from '../models/model.js';
/**
 * Test model provider that returns a predefined stream of events.
 * Useful for testing Model.streamAggregated() and other Model functionality
 * without requiring actual API calls.
 *
 * @example
 * ```typescript
 * const provider = new TestModelProvider(async function* () {
 *   yield { type: 'modelMessageStartEvent', role: 'assistant' }
 *   yield { type: 'modelContentBlockStartEvent' }
 *   yield { type: 'modelContentBlockDeltaEvent', delta: { type: 'textDelta', text: 'Hello' } }
 *   yield { type: 'modelContentBlockStopEvent' }
 *   yield { type: 'modelMessageStopEvent', stopReason: 'endTurn' }
 * })
 *
 * const message = await collectAggregated(provider.streamAggregated(messages))
 * ```
 */
export class TestModelProvider extends Model {
    eventGenerator;
    config = { modelId: 'test-model' };
    constructor(eventGenerator) {
        super();
        this.eventGenerator = eventGenerator;
    }
    setEventGenerator(eventGenerator) {
        this.eventGenerator = eventGenerator;
    }
    updateConfig(modelConfig) {
        this.config = { ...this.config, ...modelConfig };
    }
    getConfig() {
        return this.config;
    }
    async *stream(_messages, _options) {
        if (!this.eventGenerator) {
            throw new Error('Event generator not set');
        }
        yield* this.eventGenerator();
    }
}
/**
 * Helper function to collect events and result from an async generator.
 * Properly handles AsyncGenerator where the final value is returned
 * rather than yielded.
 *
 * @param generator - An async generator that yields items and returns a final result
 * @returns Object with items array (yielded values) and result (return value)
 */
export async function collectGenerator(generator) {
    const items = [];
    let done = false;
    let result;
    while (!done) {
        const { value, done: isDone } = await generator.next();
        done = isDone ?? false;
        if (!done) {
            items.push(value);
        }
        else {
            result = value;
        }
    }
    return { items, result: result };
}
/**
 * Helper function to collect all items from an async iterator.
 *
 * @param stream - An async iterable that yields items
 * @returns Array of all yielded items
 */
export async function collectIterator(stream) {
    const items = [];
    for await (const item of stream) {
        items.push(item);
    }
    return items;
}
