// @strands-agents/sdk/dist/src/__fixtures__/tool-helpers.js
/**
 * Test fixtures and helpers for Tool testing.
 * This module provides utilities for testing Tool implementations.
 */
import { ToolResultBlock } from '../types/messages.js';
import { AgentState } from '../agent/state.js';
/**
 * Helper to create a mock ToolContext for testing.
 *
 * @param toolUse - The tool use request
 * @param agentState - Optional initial agent state
 * @returns Mock ToolContext object
 */
export function createMockContext(toolUse, agentState) {
    return {
        toolUse,
        agent: {
            state: new AgentState(agentState),
            messages: [],
        },
    };
}
/**
 * Helper to create a mock tool for testing.
 *
 * @param name - The name of the mock tool
 * @param resultFn - Function that returns a ToolResultBlock or an AsyncGenerator that yields nothing and returns a ToolResultBlock
 * @returns Mock Tool object
 */
export function createMockTool(name, resultFn) {
    return {
        name,
        description: `Mock tool ${name}`,
        toolSpec: {
            name,
            description: `Mock tool ${name}`,
            inputSchema: { type: 'object', properties: {} },
        },
        // eslint-disable-next-line require-yield
        async *stream(_context) {
            const result = resultFn();
            if (typeof result === 'object' && result !== null && Symbol.asyncIterator in result) {
                // For generators that throw errors
                const gen = result;
                let done = false;
                while (!done) {
                    const { value, done: isDone } = await gen.next();
                    done = isDone ?? false;
                    if (done) {
                        return value;
                    }
                }
                // This should never be reached but TypeScript needs a return
                throw new Error('Generator ended unexpectedly');
            }
            else {
                return result;
            }
        },
    };
}
/**
 * Helper to create a simple mock tool with minimal configuration for testing.
 * This is a lighter-weight version of createMockTool for scenarios where the tool's
 * execution behavior is not relevant to the test.
 *
 * @param name - Optional name of the mock tool (defaults to a random UUID)
 * @returns Mock Tool object
 */
export function createRandomTool(name) {
    const toolName = name ?? globalThis.crypto.randomUUID();
    return createMockTool(toolName, () => new ToolResultBlock({
        toolUseId: 'test-id',
        status: 'success',
        content: [],
    }));
}
