// @strands-agents/sdk/dist/src/types/media.js
/**
 * Media and document content types for multimodal AI interactions.
 *
 * This module provides types for handling images, videos, and documents
 * with support for multiple sources (bytes, S3, URLs, files).
 */
import { TextBlock } from './messages.js';
/**
 * Cross-platform base64 encoding function that works in both browser and Node.js environments.
 */
export function encodeBase64(str) {
    if (typeof globalThis.btoa === 'function') {
        return globalThis.btoa(str);
    }
    // Node.js environment
    return globalThis.Buffer.from(str, 'binary').toString('base64');
}
/**
 * S3 location for Bedrock media and document sources.
 */
export class S3Location {
    uri;
    bucketOwner;
    constructor(data) {
        this.uri = data.uri;
        if (data.bucketOwner !== undefined) {
            this.bucketOwner = data.bucketOwner;
        }
    }
}
/**
 * Image content block.
 */
export class ImageBlock {
    /**
     * Discriminator for image content.
     */
    type = 'imageBlock';
    /**
     * Image format.
     */
    format;
    /**
     * Image source.
     */
    source;
    constructor(data) {
        this.format = data.format;
        this.source = this._convertSource(data.source);
    }
    _convertSource(source) {
        if ('bytes' in source) {
            return {
                type: 'imageSourceBytes',
                bytes: source.bytes,
            };
        }
        if ('url' in source) {
            return {
                type: 'imageSourceUrl',
                url: source.url,
            };
        }
        if ('s3Location' in source) {
            return {
                type: 'imageSourceS3Location',
                s3Location: new S3Location(source.s3Location),
            };
        }
        throw new Error('Invalid image source');
    }
}
/**
 * Video content block.
 */
export class VideoBlock {
    /**
     * Discriminator for video content.
     */
    type = 'videoBlock';
    /**
     * Video format.
     */
    format;
    /**
     * Video source.
     */
    source;
    constructor(data) {
        this.format = data.format;
        this.source = this._convertSource(data.source);
    }
    _convertSource(source) {
        if ('bytes' in source) {
            return {
                type: 'videoSourceBytes',
                bytes: source.bytes,
            };
        }
        if ('s3Location' in source) {
            return { type: 'videoSourceS3Location', s3Location: new S3Location(source.s3Location) };
        }
        throw new Error('Invalid video source');
    }
}
/**
 * Document content block.
 */
export class DocumentBlock {
    /**
     * Discriminator for document content.
     */
    type = 'documentBlock';
    /**
     * Document name.
     */
    name;
    /**
     * Document format.
     */
    format;
    /**
     * Document source.
     */
    source;
    /**
     * Citation configuration.
     */
    citations;
    /**
     * Context information for the document.
     */
    context;
    constructor(data) {
        this.name = data.name;
        this.format = data.format;
        this.source = this._convertSource(data.source);
        if (data.citations !== undefined) {
            this.citations = data.citations;
        }
        if (data.context !== undefined) {
            this.context = data.context;
        }
    }
    _convertSource(source) {
        if ('bytes' in source) {
            return {
                type: 'documentSourceBytes',
                bytes: source.bytes,
            };
        }
        if ('text' in source) {
            return {
                type: 'documentSourceText',
                text: source.text,
            };
        }
        if ('content' in source) {
            return {
                type: 'documentSourceContentBlock',
                content: source.content.map((block) => new TextBlock(block.text)),
            };
        }
        if ('s3Location' in source) {
            return {
                type: 'documentSourceS3Location',
                s3Location: new S3Location(source.s3Location),
            };
        }
        throw new Error('Invalid document source');
    }
}
