// @strands-agents/sdk/dist/src/models/streaming.js
/**
 * Event emitted when a new message starts in the stream.
 */
export class ModelMessageStartEvent {
    /**
     * Discriminator for message start events.
     */
    type = 'modelMessageStartEvent';
    /**
     * The role of the message being started.
     */
    role;
    constructor(data) {
        this.role = data.role;
    }
}
/**
 * Event emitted when a new content block starts in the stream.
 */
export class ModelContentBlockStartEvent {
    /**
     * Discriminator for content block start events.
     */
    type = 'modelContentBlockStartEvent';
    /**
     * Information about the content block being started.
     * Only present for tool use blocks.
     */
    start;
    constructor(data) {
        if (data.start !== undefined) {
            this.start = data.start;
        }
    }
}
/**
 * Event emitted when there is new content in a content block.
 */
export class ModelContentBlockDeltaEvent {
    /**
     * Discriminator for content block delta events.
     */
    type = 'modelContentBlockDeltaEvent';
    /**
     * Index of the content block being updated.
     */
    contentBlockIndex;
    /**
     * The incremental content update.
     */
    delta;
    constructor(data) {
        this.delta = data.delta;
    }
}
/**
 * Event emitted when a content block completes.
 */
export class ModelContentBlockStopEvent {
    /**
     * Discriminator for content block stop events.
     */
    type = 'modelContentBlockStopEvent';
    constructor(_data) { }
}
/**
 * Event emitted when the message completes.
 */
export class ModelMessageStopEvent {
    /**
     * Discriminator for message stop events.
     */
    type = 'modelMessageStopEvent';
    /**
     * Reason why generation stopped.
     */
    stopReason;
    /**
     * Additional provider-specific response fields.
     */
    additionalModelResponseFields;
    constructor(data) {
        this.stopReason = data.stopReason;
        if (data.additionalModelResponseFields !== undefined) {
            this.additionalModelResponseFields = data.additionalModelResponseFields;
        }
    }
}
/**
 * Event containing metadata about the stream.
 * Includes usage statistics, performance metrics, and trace information.
 */
export class ModelMetadataEvent {
    /**
     * Discriminator for metadata events.
     */
    type = 'modelMetadataEvent';
    /**
     * Token usage information.
     */
    usage;
    /**
     * Performance metrics.
     */
    metrics;
    /**
     * Trace information for observability.
     */
    trace;
    constructor(data) {
        if (data.usage !== undefined) {
            this.usage = data.usage;
        }
        if (data.metrics !== undefined) {
            this.metrics = data.metrics;
        }
        if (data.trace !== undefined) {
            this.trace = data.trace;
        }
    }
}
