// @strands-agents/sdk/dist/src/tools/mcp-tool.js
import { createErrorResult, Tool } from './tool.js';
import { JsonBlock, TextBlock, ToolResultBlock } from '../types/messages.js';
/**
 * A Tool implementation that proxies calls to a remote MCP server.
 *
 * Unlike FunctionTool, which wraps local logic, McpTool delegates execution
 * to the connected McpClient and translates the SDK's response format
 * directly into ToolResultBlocks.
 */
export class McpTool extends Tool {
    name;
    description;
    toolSpec;
    mcpClient;
    constructor(config) {
        super();
        this.name = config.name;
        this.description = config.description;
        this.toolSpec = {
            name: config.name,
            description: config.description,
            inputSchema: config.inputSchema,
        };
        this.mcpClient = config.client;
    }
    // eslint-disable-next-line require-yield
    async *stream(toolContext) {
        const { toolUseId, input } = toolContext.toolUse;
        try {
            // Input is validated by MCP Client before invocation
            const rawResult = await this.mcpClient.callTool(this, input);
            if (!this._isMcpToolResult(rawResult)) {
                throw new Error('Invalid tool result from MCP Client: missing content array');
            }
            const content = rawResult.content.map((item) => {
                if (this._isMcpTextContent(item)) {
                    return new TextBlock(item.text);
                }
                return new JsonBlock({ json: item });
            });
            if (content.length === 0) {
                content.push(new TextBlock('Tool execution completed successfully with no output.'));
            }
            return new ToolResultBlock({
                toolUseId,
                status: rawResult.isError ? 'error' : 'success',
                content,
            });
        }
        catch (error) {
            return createErrorResult(error, toolUseId);
        }
    }
    /**
     * Type Guard: Checks if value matches the expected MCP SDK result shape.
     * \{ content: unknown[]; isError?: boolean \}
     */
    _isMcpToolResult(value) {
        if (typeof value !== 'object' || value === null) {
            return false;
        }
        // Safe cast to generic record to check properties
        const record = value;
        return Array.isArray(record.content);
    }
    /**
     * Type Guard: Checks if an item is a Text content block.
     * \{ type: 'text'; text: string \}
     */
    _isMcpTextContent(value) {
        if (typeof value !== 'object' || value === null) {
            return false;
        }
        const record = value;
        return record.type === 'text' && typeof record.text === 'string';
    }
}
