// @strands-agents/sdk/dist/src/tools/tool.js
import { TextBlock, ToolResultBlock } from '../types/messages.js';
import { normalizeError } from '../errors.js';
/**
 * Event yielded during tool execution to report streaming progress.
 * Tools can yield zero or more of these events before returning the final ToolResult.
 *
 * @example
 * ```typescript
 * const streamEvent = new ToolStreamEvent({
 *   data: 'Processing step 1...'
 * })
 *
 * // Or with structured data
 * const streamEvent = new ToolStreamEvent({
 *   data: { progress: 50, message: 'Halfway complete' }
 * })
 * ```
 */
export class ToolStreamEvent {
    /**
     * Discriminator for tool stream events.
     */
    type = 'toolStreamEvent';
    /**
     * Caller-provided data for the progress update.
     * Can be any type of data the tool wants to report.
     */
    data;
    constructor(eventData) {
        if (eventData.data !== undefined) {
            this.data = eventData.data;
        }
    }
}
/**
 * Interface for tool implementations.
 * Tools are used by agents to interact with their environment and perform specific actions.
 *
 * The Tool interface provides a streaming execution model where tools can yield
 * progress events during execution before returning a final result.
 *
 * Most implementations should use FunctionTool rather than implementing this interface directly.
 */
export class Tool {
}
/**
 * Creates an error ToolResultBlock from an error object.
 * Ensures all errors are normalized to Error objects and includes the original error
 * in the ToolResultBlock for inspection by hooks, error handlers, and event loop.
 *
 * TODO: Implement consistent logging format as defined in #30
 * This error should be logged to the caller using the established logging pattern.
 *
 * @param error - The error that occurred (can be Error object or any thrown value)
 * @param toolUseId - The tool use ID for the ToolResultBlock
 * @returns A ToolResultBlock with error status, error message content, and original error object
 */
export function createErrorResult(error, toolUseId) {
    // Ensure error is an Error object (wrap non-Error values)
    const errorObject = normalizeError(error);
    return new ToolResultBlock({
        toolUseId,
        status: 'error',
        content: [new TextBlock(`Error: ${errorObject.message}`)],
        error: errorObject,
    });
}
