// @strands-agents/sdk/dist/src/hooks/events.js
/**
 * Base class for all hook events.
 * Hook events are emitted at specific points in the agent lifecycle.
 */
export class HookEvent {
    /**
     * @internal
     * Check if callbacks should be reversed for this event.
     * Used by HookRegistry for callback ordering.
     */
    _shouldReverseCallbacks() {
        return false;
    }
}
/**
 * Event triggered at the beginning of a new agent request.
 * Fired before any model inference or tool execution occurs.
 */
export class BeforeInvocationEvent extends HookEvent {
    type = 'beforeInvocationEvent';
    agent;
    constructor(data) {
        super();
        this.agent = data.agent;
    }
}
/**
 * Event triggered at the end of an agent request.
 * Fired after all processing completes, regardless of success or error.
 * Uses reverse callback ordering for proper cleanup semantics.
 */
export class AfterInvocationEvent extends HookEvent {
    type = 'afterInvocationEvent';
    agent;
    constructor(data) {
        super();
        this.agent = data.agent;
    }
    _shouldReverseCallbacks() {
        return true;
    }
}
/**
 * Event triggered when the framework adds a message to the conversation history.
 * Fired during the agent loop execution for framework-generated messages.
 * Does not fire for initial messages from AgentConfig or user input messages.
 */
export class MessageAddedEvent extends HookEvent {
    type = 'messageAddedEvent';
    agent;
    message;
    constructor(data) {
        super();
        this.agent = data.agent;
        this.message = data.message;
    }
}
/**
 * Event triggered just before a tool is executed.
 * Fired after tool lookup but before execution begins.
 */
export class BeforeToolCallEvent extends HookEvent {
    type = 'beforeToolCallEvent';
    agent;
    toolUse;
    tool;
    constructor(data) {
        super();
        this.agent = data.agent;
        this.toolUse = data.toolUse;
        this.tool = data.tool;
    }
}
/**
 * Event triggered after a tool execution completes.
 * Fired after tool execution finishes, whether successful or failed.
 * Uses reverse callback ordering for proper cleanup semantics.
 */
export class AfterToolCallEvent extends HookEvent {
    type = 'afterToolCallEvent';
    agent;
    toolUse;
    tool;
    result;
    error;
    constructor(data) {
        super();
        this.agent = data.agent;
        this.toolUse = data.toolUse;
        this.tool = data.tool;
        this.result = data.result;
        if (data.error !== undefined) {
            this.error = data.error;
        }
    }
    _shouldReverseCallbacks() {
        return true;
    }
}
/**
 * Event triggered just before the model is invoked.
 * Fired before sending messages to the model for inference.
 */
export class BeforeModelCallEvent extends HookEvent {
    type = 'beforeModelCallEvent';
    agent;
    constructor(data) {
        super();
        this.agent = data.agent;
    }
}
/**
 * Event triggered after the model invocation completes.
 * Fired after the model finishes generating a response, whether successful or failed.
 * Uses reverse callback ordering for proper cleanup semantics.
 *
 * Note: stopData may be undefined if an error occurs before the model completes.
 */
export class AfterModelCallEvent extends HookEvent {
    type = 'afterModelCallEvent';
    agent;
    stopData;
    error;
    /**
     * Optional flag that can be set by hook callbacks to request a retry of the model call.
     * Only valid when an error is present. When set to true, the agent will retry the model invocation.
     * Typically used after reducing context size in response to a ContextWindowOverflowError.
     */
    retryModelCall;
    constructor(data) {
        super();
        this.agent = data.agent;
        if (data.stopData !== undefined) {
            this.stopData = data.stopData;
        }
        if (data.error !== undefined) {
            this.error = data.error;
        }
    }
    _shouldReverseCallbacks() {
        return true;
    }
}
/**
 * Event triggered for each streaming event from the model.
 * Allows hooks to observe individual streaming events during model inference.
 * Provides read-only access to streaming events.
 *
 * Currently private pending https://github.com/strands-agents/sdk-typescript/issues/288
 */
export class ModelStreamEventHook extends HookEvent {
    type = 'modelStreamEventHook';
    agent;
    event;
    constructor(data) {
        super();
        this.agent = data.agent;
        this.event = data.event;
    }
}
/**
 * Event triggered before executing tools.
 * Fired when the model returns tool use blocks that need to be executed.
 */
export class BeforeToolsEvent extends HookEvent {
    type = 'beforeToolsEvent';
    agent;
    message;
    constructor(data) {
        super();
        this.agent = data.agent;
        this.message = data.message;
    }
}
/**
 * Event triggered after all tools complete execution.
 * Fired after tool results are collected and ready to be added to conversation.
 * Uses reverse callback ordering for proper cleanup semantics.
 */
export class AfterToolsEvent extends HookEvent {
    type = 'afterToolsEvent';
    agent;
    message;
    constructor(data) {
        super();
        this.agent = data.agent;
        this.message = data.message;
    }
    _shouldReverseCallbacks() {
        return true;
    }
}
